bin/interpret/funcfile.py

```diff
@@ -140,12 +158,12 @@ def mergeCNfinal(funcDict):
 	#		print 'SMALL SEG NEEDED MERGING'
 
 			#situations where you automatically merge with the previous segment
-			if (i == len(mergePass) - 1) or (j['chrom'] != mergePass[i+1]['chrom']) or (mergePass[i+1]['pass'] == 'no' != mergeSmall[-1]):
```

Comment:
your changes here look fine (safer, hooray! :) ) 

But something about the structure of the last conditional(s) looks weird to me from both our versions :
`(mergePass[i+1]['pass'] != 'no' and mergeSmall[-1])`
shouldn't it be something like:
`(mergePass[i+1]['pass'] != 'no' and mergeSmall[-1]['pass'] == 'no')`
?????

I might be forgetting the overall data structure I used, but I can't picture this working...the clause is trying to say "if the previous segment didn't fail FUnC and the next segment did, merge with the previous segment".

Can you test (verbosely) if the code works as written, or if it need the additional specification? If it needs the extra specification, it might need another tweak to still prevent attempting to merge with a non-existing segment...

Reply:
When I wrote the change here I tried to just replicate the same clause that you originally had without the potential for merging with a non-existent segment.
By my interpretation (based on reading through the logic, which could be wrong) the situations where you automatically merge with the previous segment are (1) if the last segment is <3 bins, (2) if the next segment is on a different chromosome, and (3) if the next segment passes FUnC (labeled as 'cnv' or 'eup') AND there's a previous segment to merge with. 
The original clause that was written allowed for two versions of the third situation to potentially be true (even though only one situation makes sense)- (1) the next segment was not CNV/euploid and there does not exist a previous segment and (2) the next segment was CNV/euploid and there exists a previous segment. 

If the clause is trying to say "if the previous segment didn't fail FUnC and the next segment did, merge with the previous segment" shouldn't it be:
(mergePass[i+1] == 'no' and mergeSmall[-1] != 'no')

However, shouldn't the logic be opposite that ^^ where you don't want to falsely extend the boundaries/size of a CNV if you're not confident, so you should actually merge with the previous segment if it didn't pass FUnC and the next segment does pass FUnC (which is consistent with (mergePass[i+1]['pass'] != 'no' and mergeSmall[-1]['pass'] == 'no') as you also suggested above)
[am I potentially just confusing myself now?...]

Now that I think about it, checking for existence of previous/next segments is a little bit unnecessary because the checks for whether the segment is the first or last segment examined should be the only cases in which a previous or next segment doesn't exist.

Reply:
ughhhhhh, this logic really got confusing...I had a hard time figuring out alternates that would read more simply.

My first concern is having no idea off the top of my head what 
>(mergePass[i+1]['pass'] != 'no' and mergeSmall[-1])
is going to evaluate regardless of the specific entry, since mergeSmall[-1] should be a dictionary...so I'd guess this check will ALWAYS consider mergeSmall[-1] to be True, which is meaningless. I swear I troubleshot this code for a few hours, but I also don't want to assume I didn't miss an issue like this (my test set was just dev036).

Second, I think you are interpreting my meaning correctly -- though I'm pretty sure FUnC only assesses CNVs, so no euploid segments will be labeled 'pass'. But I'm pretty sure it was a conscious choice to merge small, non-passing segments with a CNV segment that did pass. This was partly influenced by believing that small segments won't change much for whatever they're merged with--but I'll admit that's not a guarantee. I also know this choice was influenced by looking at data and seeing it made more sense to merge with reliable segments...but, again, it was a limited test case and I don't remember the details.

The only way to really be confident in the behavior is to look at more test cases in detail...how often does this decision come up, and is there a clear choice that looks most appropriate? This particular situation probably deserves more nuanced options  - such as merging with the 'pass' CNVs when they are large, but not small. 

Most of the time, this part of the code won't be activated, and when it does it will likely have little impact on the CNV call output. So I'm not sure how much effort it warrants. Do you have an opinion?

Reply:
`(mergePass[i+1]['pass'] != 'no' and mergeSmall[-1]) `
1. Checks to make sure that there is a previous segment (mergeSmall[-1]) (which is a redundant check, because as long as it's not the first segment being assessed, there will be a previous segment), and
2. checks to make sure that the next segment passed FUnC
The 'pass' field has three different options: 'no' (for didn't pass FUnC), 'eup', and 'cnv', and at this point it hasn't only selected the CNVs to look at more carefully, so I believe there will be euploid segments labeled as passing FUnC. 

Now that I'm re-reading this again, I think we're overthinking this (and that I have some logic slightly backwards) and that the places to automatically merge are:
1) with a previous segment if the next segment doesn't pass FUnC ['pass']=='no'
2) with the next segment if the previous segment doesn't pass FUnC ['pass']=='no'
because the small segment needs to be merged with a similar adjacent segment (whether that be euploid or cnv), and if it's not either of these cases, the final else clause figures out which segment (previous or next) to merge with based on CN. 

So this would actually give us
```
#situations where you automatically merge with the previous segment
if (i==len(mergePass)-1) or (j['chrom'] != mergePass[i+1]['chrom']) or (mergePass[i+1]['pass']=='no' and mergeSmall[-1])
#situations where you automatically merge with the next segment
elif (i==0) or (j['chrom'] != mergeSmall[-1]['chrom']) or (mergeSmall[-1]['pass'] == 'no' and mergePass[i+1])
else #conditions where you need to figure out which segment to merge with 
```

Reply:
hah! we probably are overthinking it.

I agree with your #1 and #2 -- auto merge w/ previous segment if the next one has pass==no and auto merge w/ next segment if previous one has pass==no.

The else clause might deserve some extra logic-ing (or I might still be overthinking)
A: if both surrounding segments failed -- yes, just merge based on CN similarity
B: if both surrounding segments pass and both are euploid -- it doesn't matter who is selected b/c all 3 segments will ultimately be merged together
C: if both surrounding segments pass and both are CNVs -- again, merge based on CN
D: if both surrounding segments pass and one is a CNV and one is euploid -- it might be safer to give some degree of preference to merging with the euploid segment. Not sure if that would be better if you always merge with the euploid one, or if the CN difference from the CNV segment must be x% smaller than the CN difference from euploid. 

For D, if will probably be obvious which to use from looking at a few data examples. If it seems like the fail segments are often some weird edge effects of a continuous CNV, it should be possible for them to get merged in. If they seem unrelated most of the time, just merge all of these with euploid.